### PR TITLE
WIP: POST requests with fetch

### PR DESCRIPTION
**Pull request by Ashniu123**
WIP!

Code is not complete(not even compiling) right now but awaiting confirmation that progress is being made in the right direction.

To close issue [#1133](https://github.com/denoland/deno/issues/1133).

### CODE REVIEW

File: src/ops.rs [hunk 1]
```diff
@@ -408,8 +409,19 @@ fn op_fetch(
   let url = url.parse::<hyper::Uri>().unwrap();
   let client = http_util::get_client();
 
+  let mut req = hyper::Request::new(hyper::Body::from(&*data));
+  *req.method_mut() = match method {
+    "POST"=> hyper::Method::POST,
+    _ => hyper::Method::GET
```
[the remark this answers is not in the document]
Comment (Ashniu123) [reply src/ops.rs]: Alright. Will do that.

**Comment by Ashniu123** [comment]
I have a doubt [here](https://github.com/denoland/deno/blob/3422b7507017636f0f1f143d5f126f2bf5130bcc/src/ops.rs#L44). If I add `method` as `str` I would have to add it to all `op_*` functions. Is that fine or is there a better way to do it?

**Comment by hayd** [comment]
Look at how the other op functions take arguments. You don't need to modify any other op functions.

**Comment by ry** [comment]
@Ashniu123 It doesn't need a 4th parameter, `method` should be passed via the message (aka `base`). See the other examples for how to extract it.

File: js/fetch.ts [hunk 2]
```diff
@@ -198,12 +198,17 @@ export async function fetch(
 ): Promise<domTypes.Response> {
   const url = input as string;
   log("dispatch FETCH_REQ", url);
+  const method = init && init.method ? (init.method as string) : "GET";
```
Comment (hayd) [inline js/fetch.ts]: 👍 

I think you may need to implement Request for this, see #191. Note that .headers is an attribute of Request so you'll be able to send that too. (Maybe .body too?)

Comment (hayd) [reply js/fetch.ts]: Maybe see also https://github.com/superfly/fly.rs/blob/master/v8env/src/request.ts

File: src/ops.rs [hunk 3]
```diff
@@ -400,6 +400,7 @@ fn op_fetch(
   let cmd_id = base.cmd_id();
   let id = inner.id();
   let url = inner.url().unwrap();
+  let method = inner.method().unwrap_or("GET");
```
Comment (bartlomieju) [inline src/ops.rs]: I think `unwrap_or` is not needed, `method` is always passed in message.

File: src/ops.rs [hunk 4]
```diff
@@ -408,8 +409,20 @@ fn op_fetch(
...
+  let mut req = hyper::Request::new(hyper::Body::from(&*data));
+  *req.method_mut() = match method {
+    "POST" => hyper::Method::POST,
+    "GET" => hyper::Method::GET,
+    _ => panic!("Invalid method"),
+  };
+  *req.uri_mut() = url.clone();
+  req.headers_mut().insert(
+    hyper::header::CONTENT_TYPE,
+    hyper::header::HeaderValue::from_static("application/json"),
```
Comment (bartlomieju) [inline src/ops.rs]: Content type shouldn't be hardcoded. It should be taken from headers, which at the moment are not passed to Rust.

File: src/ops.rs (hunk 3 above)
Comment (Ashniu123) [reply src/ops.rs]: Will do that. Thanks!

File: js/fetch.ts [hunk 5]
```diff
@@ -198,12 +198,20 @@ export async function fetch(
 ): Promise<domTypes.Response> {
   const url = input as string;
   log("dispatch FETCH_REQ", url);
-  const method = init && init.method ? (init.method as string) : "GET";
+
+  let method = "GET";
+  if (typeof input === "string" && init && init.method) {
```
[the remark this answers is not in the document]
Comment (Ashniu123) [reply js/fetch.ts]: Oh! Thanks!

File: src/ops.rs (hunk 4 above)
Comment (Ashniu123) [reply src/ops.rs]: Yeah I know but it was just a quick implementation before I added headers. Will be fixed in next few commits.

File: js/fetch.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (hayd) [reply js/fetch.ts]: that'll default the GET for you too

File: js/body_mixin.ts [hunk 6]
```diff
@@ -0,0 +1,230 @@
+/** @module fly
```
Comment (kitsonk) [inline js/body_mixin.ts]: We do not use this annotation in other modules.  It adds no value to the intellisense for TypeScript, so I wouldn't recommend using it.  Also what is `fly` anyways?

File: js/body_mixin.ts [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+  FormData,
+  Body,
+  ReadableStream,
+  ReadableStreamReader,
+  BodyInit
+} from "./dom_types";
+import { DenoBlob } from "./blob";
+import DenoFormData from "./form_data";
+
+// interface ReadableStreamController {
```
Comment (kitsonk) [inline js/body_mixin.ts]: Why is this commented out?

File: js/body_mixin.ts [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    } else if (!this.bodySource) {
+      return new ArrayBuffer(0);
+    }
+    throw new Error(
+      `Body type not yet implemented: ${this.bodySource.constructor.name}`
+    );
+  }
+}
+
+/** @hidden */
```
Comment (kitsonk) [inline js/body_mixin.ts]: Again, JSDoc we don't have elsewhere.  It is a non exported function anyways, I am not sure what value there is in this annotation.

File: js/body_mixin.ts [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+function validateBodyType(owner: any, bodySource: any) {
```
Comment (kitsonk) [inline js/body_mixin.ts]: tslint won't allow this `any`, I don't this has been run through the linter.  Can't we be more specific here.

File: js/body_mixin.ts (hunk 6 above)
Comment (kitsonk) [reply js/body_mixin.ts]: Why is this module named `body_mixin`?  It is not a mixin in any sense.  It exports a class named `DenoBody`, which implements `domTypes.Body`.  It should just be named `body.ts`

File: js/body_mixin.ts [hunk 10]
```diff
@@ -0,0 +1,230 @@
+/** @module fly
+ */
+import { parse as queryParse } from "query-string";
+import {
```
Comment (kitsonk) [inline js/body_mixin.ts]: in order to build the library properly, I believe this needs to be a namespaced import:

```ts
import * as domTypes from "./dom_types";
```

File: js/body_mixin.ts [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+//   close(): void
+// }
+
+export type BodySource =
+  | Blob
+  | BufferSource
+  | FormData
+  | URLSearchParams
+  | ReadableStream
+  | String;
```
Comment (kitsonk) [inline js/body_mixin.ts]: `string` never `String`

File: js/body_mixin.ts [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+export default class DenoBody implements Body {
+  protected bodySource: BodyInit;
+  protected stream: ReadableStream | null;
+
+  constructor(obj: BodyInit) {
```
Comment (kitsonk) [inline js/body_mixin.ts]: `init` would be better than `obj`.  `obj` is non-obvious.

File: js/body_mixin.ts [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+    validateBodyType(this, obj);
+    this.bodySource = obj;
+    this.stream = null;
```
Comment (kitsonk) [inline js/body_mixin.ts]: better to do this as the property initialiser above:

```
protected stream: ReadableStream | null = null;
```

File: js/body_mixin.ts [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+  }
+
+  get body(): ReadableStream | null {
```
Comment (kitsonk) [inline js/body_mixin.ts]: how can `null` be returned from this?  I can't see how from the code below.

File: js/body_mixin.ts [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+  }
+
+  get staticBody(): Uint8Array {
+    if (this.bodySource instanceof Uint8Array) return this.bodySource;
+    else if (typeof this.bodySource === "string")
+      return new TextEncoder().encode(this.bodySource);
+    else throw new TypeError("body is not static");
+  }
+
+  set body(value: ReadableStream) {
```
Comment (kitsonk) [inline js/body_mixin.ts]: the get/set should be paired together, so they are easier to find in the code

File: js/body_mixin.ts [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+          if (typeof value === "string") {
+            parts.push(encoder.encode(value));
+          } else if (value instanceof ArrayBuffer) {
+            parts.push(new Uint8Array(value));
+          } else if (value instanceof Uint8Array) {
+            parts.push(value);
+          } else if (!value) {
+            // noop for undefined
+          } else {
+            console.log("unhandled type on stream read:", value);
```
Comment (kitsonk) [inline js/body_mixin.ts]: we should just reject... why are we writing to the console?

File: js/body_mixin.ts [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+export function bufferFromStream(
+  stream: ReadableStreamReader
+): Promise<ArrayBuffer> {
+  return new Promise((resolve, reject) => {
+    let parts: Array<Uint8Array> = [];
+    let encoder = new TextEncoder();
+    // recurse
+    (function pump() {
```
Comment (kitsonk) [inline js/body_mixin.ts]: I am not sure I like this IIFE at all... I need to take a look at the `ReadableStreamReader`, but it feels like this is an async iterator and could be handled much cleaner.  What is the justification for using an IIFE here?

File: js/body_mixin.ts [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+          reject(err);
+        });
+    })();
+  });
+}
+
+/** @hidden */
+function concatenate(...arrays: Uint8Array[]): ArrayBuffer {
+  let totalLength = 0;
+  for (let arr of arrays) {
```
Comment (kitsonk) [inline js/body_mixin.ts]:
```suggestion
  for (const arr of arrays) {
```

File: js/body_mixin.ts [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+    totalLength += arr.length;
+  }
+  let result = new Uint8Array(totalLength);
+  let offset = 0;
+  for (let arr of arrays) {
```
Comment (kitsonk) [inline js/body_mixin.ts]:
```suggestion
  for (let const of arrays) {
```

File: js/body_mixin.ts [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+    result.set(arr, offset);
+    offset += arr.length;
+  }
+  return <ArrayBuffer>result.buffer;
```
Comment (kitsonk) [inline js/body_mixin.ts]: don't coerce this way... 
```suggestion
  return result.buffer as ArrayBuffer;
```

But I am not sure why you even need the coercion here.

File: js/body_mixin.ts (hunk 6 above)
Comment (Ashniu123) [reply js/body_mixin.ts]: Okay. I'll fix that.

File: tsconfig.json [hunk 21]
```diff
@@ -19,6 +19,7 @@
     "resolveJsonModule": true,
     "sourceMap": true,
     "strict": true,
+    "strictPropertyInitialization": false,
```
Comment (kitsonk) [inline tsconfig.json]: Please do not do this.  We should not be working around bad code by making TypeScript less strict.

File: src/ops.rs (hunk 3 above)
Comment (bartlomieju) [reply src/ops.rs]: I think it's not yet resolved

File: tsconfig.json (hunk 21 above)
Comment (Ashniu123) [reply tsconfig.json]: I agree actually but what happened is that last time, it never showed the errors `Property 'referrerPolicy' has no initializer and is not definitely assigned in the constructor.` everywhere until now. So this was the only workaround. I'll revert it. Sorry!

Comment (Ashniu123) [reply tsconfig.json]: Also please review the remaining stuff, how good (or bad) is it? Also I had an issue with `headers` in `fetch.ts`'s `fetch` method. If you could please look into it.

**Comment by ry** [comment]
I appreciate the attempt - but I'd rather do this another way
closed in favor of #1188

**Pull request closed without merging**